src/io/fasta.rs

```diff
@@ -252,12 +361,73 @@ struct IndexRecord {
 
 
 /// A sequence record returned by the FASTA index.
+#[derive(Debug, PartialEq)]
 pub struct Sequence {
     pub name: String,
     pub len: u64,
 }
 
 
+pub struct IndexedReaderIterator<'a, R: io::Read + io::Seek + 'a> {
+    reader: &'a mut IndexedReader<R>,
+    record: IndexRecord,
+    bases_left: u64,
+    line_offset: u64,
+    buf: Vec<u8>,
+    buf_idx: usize,
+    buf_len: usize,
+}
+
+
+impl<'a, R: io::Read + io::Seek + 'a> IndexedReaderIterator<'a, R> {
+    fn fill_buffer(&mut self) -> io::Result<()> {
+        assert!(self.bases_left > 0);
+
+        while self.buf_idx >= self.buf_len {
```

Comment:
You are right that it is not beneficial to read multiple lines (I did try it), but the purpose of that loop is to ensure that we have actually read some bases. A call to IndexedReader::read_line may end up only reading whitespace, if the current position is at the start of or inside of whitespace, as the function only reads till the end of the current line. The loop terminates as soon as one or more bases have been read into the buffer, to satisfy the assumption in next().